### PR TITLE
add EnsureWithSuccess

### PR DESCRIPTION
**Pull request by xlc**
I would also like to replace all `EnsureRootWithSuccess` to use `EnsureWithSuccess` but that doesn't need to happen in this PR.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2080485

File: frame/system/src/lib.rs [hunk 1]
```diff
@@ -816,6 +816,27 @@ impl<
...
+	sp_std::marker::PhantomData<(Ensure, AccountId, Success)>,
+);
+impl<
+		O: Into<Result<RawOrigin<AccountId>, O>> + From<RawOrigin<AccountId>>,
+		Ensure: EnsureOrigin<O>,
+		AccountId,
+		Success: TypedGet,
+	> EnsureOrigin<O> for EnsureWithSuccess<Ensure, AccountId, Success>
+{
+	type Success = Success::Type;
```
Comment (bkchr) [inline frame/system/src/lib.rs:829]:
```suggestion
	type Success = Success::Type;

```

File: frame/system/src/lib.rs [hunk 2]
```diff
@@ -816,6 +816,27 @@ impl<
 	}
 }
 
+pub struct EnsureWithSuccess<Ensure, AccountId, Success>(
+	sp_std::marker::PhantomData<(Ensure, AccountId, Success)>,
+);
```
Comment (bkchr) [inline frame/system/src/lib.rs:821]:
```suggestion
pub struct EnsureWithSuccess<Ensure, AccountId, Success>(
	sp_std::marker::PhantomData<(Ensure, AccountId, Success)>,
);

```

Comment (bkchr) [reply frame/system/src/lib.rs:821]: Missing docs.

**Review by bkchr — APPROVED** [review]
If you add docs, it should be fine.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2081452

**Review by lexnv — APPROVED** [review]

**Comment by bkchr** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2082838

**Review by shawntabrizi — APPROVED** [review]

**Review by ggwpez — APPROVED** [review]

**Comment by ggwpez** [comment]
bot merge

**Pull request merged**